### DISCUSSION TITLE
単体テストの考え方/使い方 4.2〜4.4

### DISCUSSION THREAD

**Discussion by dameleon** [atman-inc/training#97] [General]
- 4.2: 退行(regression)に対する保護とリファクタリングへの耐性との関係
- 4.3: 迅速なフィードバックと保守のしやすさ
- 4.4: 理想的なテストの探求
    - 行けるとこまで。はみ出たら次週。

上記辺りののディスカッション。
不明点、相談点、確認点、勉強になったポイントなどは各自ディスカッションのコメントをお願いします。

**Comment by y-kaneda** [comment]
## 4.2 退行(regressionに対する保護とリファクタリングへの耐性との関係)

### 4.2.1 テストの正確性を最大限にすること
- コードの正しさとテストの関係
  - テストが成功でコードも正しい -> 真陰性
  - テストが成功でコードが間違い -> 偽陰性
    - 本来テストが失敗するべき => 退行に対する保護が無い
    - テストをすることでどのくらいのバグを検出できるのか
    - 見つけられないバグがある
  - テストが失敗でコードが正しい -> 偽陽性
    - 本来テストが成功するべき => リファクタリングへの耐性が無い
    - テストをすることでバグがないことをどのくらい示せるのか
    - 嘘の警告の発生
  - テストが失敗でコードが間違い => 真陽性

### 4.2.2 偽陽性(false positive) と偽陰性(false negative)の重要性の違い
- プロジェクト初期はリファクタリングをする必要性がないので偽陽性はそこまで問題にならない
  - 機能バグのほうが注視されやすく、リファクタリングへの耐性が必要になるほどの規模・年数にならない可能性もあり、軽視されがちな面もある

## 4.3 迅速なフィードバックと保守のしやすさ
- 迅速なフィードバック
  - 速やかにテストを行えるようになるとフィードバックから改善までの時間が短くなる
    - _感想:実行時間に加えて手元でコマンド一発でカジュアルにテスト実行できるとかの話もありそう？_
  - バグが持ち込まれたらすぐに検出できるようになる
    - _感想:テスト・ケースが充実してるとかの話？_
  - _感想:テスト(実行速度)が遅くて実施回数が減るとかある？CICD無し前提？_
- 保守のしやすさ
  - テスト・ケースを理解することの難しさに影響するもの
    - テスト・コードの量
    - プロダクション・コードと同じく、コードの品質
  - テストを行うことの難しさに影響するもの
    - テストのためのプロセス外依存の量・複雑さ

## 4.4 理想的なテストの探究
- 4本の柱のどれか一つでも欠けていたらそのテストの価値は0
- テスト・コードも**負債**であることを理解し、価値のないたくさんのケースよりも価値のあるケースが必要な分だけのほうが有用

### 4.4.1 理想的なテストを作成することは可能なのか？
- 退行に対する保護・リファクタリングへの耐性・迅速なフィードバックの3本は互いに排反

### 4.4.2 極端な例＃1: E2E(End-to-End)テスト
- 多くのプロダクションコードが実行されるため、退行に対する保護を十分に備えている
- 嘘の警告を発する偽陽性(false positive)も持ち込まれにくいため、リファクタリングへの耐性も備えている
  - ユーザー視点での(一番外側の)結果を観測するため
- テスト実行に時間がかかり、迅速なフィードバックは得られない

### 4.4.3 極端な例＃2: 取るに足らないテスト
- 実行速度は速いし、プロダクションコードが変わったら即失敗するが、意味としての振る舞いの結果を見ていないので退行に対する保護はない
  - _感想:滅べ_

### 4.4.4 極端な例＃3: 壊れやすいテスト
- 実行時間が短く、退行を見つけることに優れてはいても多くの偽陽性を持ち込んでしまうテスト
- プロダクションコードに近寄り過ぎていて、コードの変更に必要以上に過敏な反応を示してしまう

### 4.4.5 理想的なテストに関する結論
- 排反する3本の柱をどうトレード・オフするか
- リファクタリングへの耐性は最大にし、退行に対する保護と迅速なフィードバックをバランスさせる
  - リファクタリングへの耐性は基本的に0 or 1になる

**Comment by nmurakami0** [comment]
4.2

- テストが成功するか(negative)、失敗(positive)するかと、実際の振る舞いが正しいか否かで分類
  - リファクタリングへの耐性があると、false positiveを抑制出来る
  - regressionに対する保護がなされていると、false negativeを抑制できる
- SN比: テストの正確性 = 信号(検出されたバグの数) / ノイズ(嘘の警告が発せられた数)
- プロジェクトの初期段階だと、false negativeほどfalse positiveは問題にならない
  - リファクタリングしない限りは、通っていたテストが通らなくなることはないため

上記説明が若干わかりづらいが、テストが全部通っている状態があり、そこからコードを足したり修正したりした場合にどうかという話

4.3

- 残りの柱
  - 迅速なフィードバック
    - テストの実行速度
  - 保守のしやすさ
    - テストケースを理解することがどれくらい難しいか
    - テストを行うのがどれくらい難しいか

E2Eの導入がなかなかうまくいかないのは、このあたりにあるのはそうだと思われる。
実際どうすれば運用が回るか。。。

4.4

- テストケースの価値は、3つの柱の掛け合わせ+保守のしやすさで決まる
- 単体テストのケースを作成する際、どの柱を優先し、どの柱を犠牲にするかを決断しなくてはならない
  - ただし、リファクタリングへの耐性を犠牲にすることは出来ない

現代ではCI環境が容易に構築出来るので、結構バランスよく作成出来る気がする。
ただ、プロジェクトが大きくなってテストケースが膨大になったときに、どれを重視するかは出てきそう。

regressionが起きたときにテストケースに付け足すことはままあるが、振る舞いのテストという意味では、
本当はケースを全体的に見直した方が良いかもしれない。

**Comment by koushisa** [comment]
## 4.2

- テストや検査、統計学の分野で利用されるユビキタス言語の解説っぽい
- `第二種過誤(Type Ⅱ Error`とか初めてきいた
- 「多くのプロジェクトはリファクタリングへの耐性を必要とするくらいの大きな規模になる前に開発が終わってしまう」
  - 崖から落ちながら飛行機を作るという例え話も似たようなものか
  - プロジェクトの立ち上げ期はとにかく必死
  - 機能単位の話であればあまり当てはまらないかもしれない
    - 例えば自分は5日の見積もりであれば2日で実装して、1日休んで4日目にまっさらな頭でセルフレビューしてリファクタすることが多い

## 4.4 理想的なテスト

- テスト手法に対してどこがトレードオフかを示している
  - すべてがトレードオフであるのはデザインも設計も同じ
  - 絶対的な 「理想的」はないので、最終的には各自の脳みそのフィルタを通した経験則からくる「相対的」にはなる
- E2Eと取るに足らないテストのバランスは難しい
  - E2Eで全てカバーするのは組合せ爆発を起こす
  - (話に出てこないが)開発者視点でロジカルな部分のテストをしたいときのinteraction testに最近注目しているので後段でまとめたい
- FE視点でdiscussionsに現時点の考えを発散した
  - https://github.com/atman-inc/training/discussions/96#discussioncomment-5549196

**Comment by myamamoto111** [comment]
## 4.2 退行(regression)に対する保護とリファクタリングへの耐性との関係

### 4.2.1 テストの正確性を最大限にすること

- 真陰性
  - テスト成功、振る舞い正しい　OK
- 真陽性
  - テスト失敗、振る舞い間違い　OK
- 偽陰性
  - テスト成功、振る舞い間違い　NG
  - 第二種過誤（Type Ⅱ error）
  - 抑えるために、退行に対する保護
- 偽陽性
  - テスト失敗、振る舞い正しい　NG
  - 第一種過誤（Type Ⅰ error）
  - 抑えるために、リファクタリングへの耐性

- 信号対ノイズ比
  - 被分子：信号を増やす
    - より多くの退行（バグ）を見つけられるように
  - 除数（分母）：ノイズを減らす
    - テストが嘘の警告を発することを減らす


### 4.2.2 偽陽性と偽陰性の重要性の違い

偽陽性はプロジェクトの開始時期にはそれほど大きな影響を与えないが、プロジェクトが成長すると、偽陽性は大きな問題となり、その重要性は偽陰性と同じくらいになる

## 4.3 迅速なフィードバックと保守のしやすさ
- テストケースを理解すること
- テストを行うこと

## 4.4 理想的なテストの探求
### 4.4.1
理想的なテストとは、良い単体テストを構成する４本の柱を全て完全に備えたテストケースのこと
しかしながら不可能
ただ、どの柱も可能な限り備えるように努める

### 4.4.2
E2Eテスト
- 退行に対する保護とリファクタリングへの耐性は十分に備わっているが
- 迅速なフィードバックは全く備わっていない

Puppeteer, Playwriteを試してみたが、
画面を表示しながら人間が見れる速度で正常に実行されているかを試すには良い
スクリーンショットや録画機能があるので、エビデンスを求められる場合には良い
ただ、初期段階やCIに組み込むのは難しい。（環境整備が大変）
あくまで、最終的なシステムテストとして、切り離して実施するのには良いと思う。

### 4.4.3
取るに足らないテスト
- リファクタリングへの耐性と迅速なフィードバックを十分に備えてはいるが
- 退行に対する保護は全く備えていない

### 4.4.4
壊れやすいテスト
- 退行に対する保護と迅速なフィードバックは備わってはいるが
- リファクタリングへの耐性は欠落している

### 4.4.5
理想的なテストに関する結論
- 3本の柱すべてを最大限に備えた理想的な単体テストを作成することは不可能
- どのようなトレードオフを図るか！？
  - 最善は保守のしやすさとリファクタリングへの耐性を最大限に備えたもの
  - 退行に対する保護と迅速なフィードバックの２本の柱のどちらを優先するかのバランス調整
- CAP定理
  - 一貫性（Consistency）
  - 可用性（Availability）
  - 分断性（Partition-tolerance）

トレードオフのバランス調整が大事
